File: Acwing/First_Level_0_Basic_Algorithm/Second_DataStructure/ds_143_maximum_xor_pair/Main.cpp
#include<iostream>
using namespace std;

const int N = 100010;

int q[3100010][2] , idx , a[N];

void insert(int s){
    int p = 0;
    for(int i = 30 ; i >= 0 ; --i){
        int u = s >> i & 1;
        if(!q[p][u]) q[p][u] = ++idx;
        p = q[p][u];
    }
}

int query(int s){
    int p = 0 , res = 0;
    for(int i = 30; i >= 0 ; --i){
        int u = s >> i & 1;
        if(q[p][!u]){
            res += 1 << i;
            p = q[p][!u];
        }else p = q[p][u];
    }
    return res;
}

int main(){
    int n;
    scanf("%d" , &n);
    for(int i = 0 ; i < n ; ++i){
        scanf("%d" , &a[i]);
        insert(a[i]);
    }
    int ans = 0;
    for(int i = 0 ; i < n ; ++i){
        ans = max(ans , query(a[i]));
    }
    cout << ans << endl;
}

File: Acwing/First_Level_0_Basic_Algorithm/Second_DataStructure/ds_143_maximum_xor_pair/Main.java
package First_Level_0_Basic_Algorithm.Second_DataStructure.ds_143_maximum_xor_pair;



import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    static final int N = 100010;
    static final int M = 310010;
    static int [][] p = new int[M][2];
    static String [] a;
    static int n , idx;

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.valueOf(br.readLine());
        a = br.readLine().split(" ");
        for(int i = 0; i < n ; ++i){
            insert(Integer.valueOf(a[i]));
        }

        int ans = 0;
        for(int i = 0 ; i < n ; ++i){
            ans = Math.max(ans, Integer.valueOf(a[i]));
        }
        System.out.println(ans);
        br.close();    
    }

    public static void insert(int x){
        int t = 0;
        for(int i = 30 ; i >= 0 ; --i){
            int u = x >> i & 1;
            if(p[t][u] <= 0) p[t][u] = ++idx;
            t = p[t][u];
        }
    }

    public static int query(int x){
        int t = 0 , res = 0;
        for(int i = 30 ; i >= 0 ; --i){
            int u = x >> 30 & 1;
            if(p[t][~u] > 0){
                res += 1 << i;
                t = p[t][~u];
            }else t = p[t][u];
        }
        return res;
    }
}
